Add slide to show the Mockall crate in action (#1532)
google/comprehensive-rust: src/testing/mocking.md

---
minutes: 10
---

# Mocking

If you want to do mocking, we recommend [Mockall]. You need to refactor your
code to use traits, which you can then quickly mock:

```rust,ignore
{{#include mockall.rs}}
```

[Mockall]: https://docs.rs/mockall/

<details>

- Mockall is not part of the Rust Playground, so you need to run this example in
  a local environment. Use `cargo add mockall` to quickly add Mockall to an
  existing Cargo project.

- Mockall has a lot more functionality. In particular, you can set up
  expectations which depend on the arguments passed:

    ```rust,ignore
    let mut mock_cat = MockPet::new();
    mock_cat
        .expect_is_hungry()
        .with(mockall::predicate::gt(Duration::from_secs(3 * 3600)))
        .return_const(true);
    mock_cat
        .expect_is_hungry()
        .return_const(true);
    assert_eq!(mock_cat.is_hungry(Duration::from_secs(1 * 3600)), false);
    assert_eq!(mock_cat.is_hungry(Duration::from_secs(5 * 3600)), true);
    ```

- You can use `.times(n)` to limit the number of times a mock method can be
  called to `n` --- the mock will automatically panic when dropped if this isn't
  satisfied.

> - Mockall is available for use in AOSP.

Author: This is not strictly true today, but it should be true by ~next week.

---

> minutes: 10

Reviewer: I don't want to litigate down to the minute, IMHO this slide should be short: "If it's mocking you need, there's a great crate for that which is used in Android, called mockall. Here's an example, which creates a MockPet type. Find more at docs.rs when you need it"

Author: Yeah, maybe 10 minutes is overestimating it here. I really have to schedule a class myself to understand how much time to spend on each slide.

Let change it to 5 minutes as a start?

```suggestion
minutes: 5
```

---

# Mocking

> If you want to do mocking, we recommend [Mockall]. You need to refactor your

Reviewer: ```suggestion
If you want to do mocking, AOSP recommends [Mockall], although there are several other crates available. You need to refactor your
```
..or something to that effect. I think the existing text implies "mockall is how people do test mocks in Rust", and that's not accurate.

Author: I did a survey and from what I can tell, Mockall is vastly more used than any other mocking library right now:

|                                               | Released | Updated | Dependants | Downloads/month |
|-----------------------------------------------|---------:|--------:|-----------:|----------------:|
| [Mockall](http://lib.rs/crates/mockall)       |     2019 |  _2023_ |        295 |              1M |
| [Mockers](http://lib.rs/crates/mockers)       |     2016 |  _2022_ |          3 |             560 |
| [Simulacrum](http://lib.rs/crates/simulacrum) |     2017 |    2018 |          1 |              6k |
| [Pseudo](http://lib.rs/crates/pseudo)         |     2017 |    2018 |          4 |             653 |
| [Double](http://lib.rs/crates/double)         |     2017 |    2019 |          1 |             208 |
| [Mockiato](http://lib.rs/crates/mockiato)     |     2019 |    2019 |          1 |              2k |
| [Mock-it](http://lib.rs/crates/mock-it)       |     2018 |  _2022_ |          2 |             406 |

Most libraries haven't been updated in a few years and most of them have very few reverse dependencies.

Of course Mockall isn't the end all of mocking, so a new library could be made which is even better. But right it's the one we will recommend to our engineers in Android.

At the end of the day, the course here is written for Android engineers. I've tried my best to take out the Android-specific parts and make as much as it generic. I think the advise here is also generic enough for this section.

However, let us add a note on the slide (and perhaps somewhere in the beginning) to say that the material here is an _opinionated_ guide to Rust. It tried to be fair and generic, but it is sponsored by Google, and it will recommend tools useful today to Android and Google engineers. Does that sound fair?

Reviewer: Ah, thanks for that analysis of the libraries! A few search results suggested a much wider range of options. Given that information, this slide makes sense.

Author: Thanks! I will add a note explaining that the recommendations on the slide is current at of now and apply directly to Android. They might also apply to others, but as you point out, there _are_ other options out there!

code to use traits, which you can then quickly mock:

```rust,ignore
{{#include mockall.rs}}
```

[Mockall]: https://docs.rs/mockall/

<details>

- Note that *mocking is controversial*: mocks allow you to completely isolate a
  test from its dependencies. The immediate result is faster and more stable
  test execution. On the other hand, the mocks can be configured wrongly and
  return output different from what the real dependencies would do.

  If at all possible, it is recommended that you use the real dependencies. As
  an example, many databases allow you to configure an in-memory backend. This
  means that you get the correct behavior in your tests, plus they are fast and
  will automatically clean up after themselves.

  Similarly, many web frameworks allow you to start an in-process server which
  binds to a random port on `localhost`. Always prefer this over mocking away

>   the framework since it helps you test your code in the real environment.

Reviewer: This seems out of context here -- if students are asking how to use mocks in Rust, they are probably more focused on the "in Rust" part and do not want to be told that they shouldn't be using mocks.

Reviewer: I still want to keep it short, though, just in the interest of overall course length. I think this lengthy speaker note encourages a digression into the relative merits of various testing strategies, which will probably generate questions and comments from students, and suddenly we've been on this slide for 15 minutes.

I agree that an article would be great, but maybe this could be moved to "More to Explore"?

At any rate, LGTM.

Author: > I still want to keep it short, though, just in the interest of overall course length.

Fair point! I was not thinking that the instructor would go through all the pros and cons — I guess I was writing this bullet point more for the people who will read the slide by themselves.

The problem of multiple personas reading the slides is tricky... as we've discussed before :smile: It just occurred to me that we could create more sections in the notes (and first rename them to "Notes"):

```markdown
<details>

### Speaker

- Make sure to mention X and Y.
- Try removing `x`, note the compilation error.

### Participants

- The concept of Foo was originally introduced in C++ as Bar. See Wikipedia.
- If you want to explore more, then try ...

</details>
```

Basically an extension of the "More to Explore" headings we've added a few places.

- Mockall is not part of the Rust Playground, so you need to run this example in
  a local environment. Use `cargo add mockall` to quickly add Mockall to an
  existing Cargo project.

> - Mockall has a lot more functionality. In particular, you can set up
>   expectations which depend on the arguments passed:
> 
>     ```rust,ignore
>     let mut mock_cat = MockPet::new();
>     mock_cat
>         .expect_is_hungry()
>         .with(mockall::predicate::gt(Duration::from_secs(3 * 3600)))
>         .return_const(true);
>     mock_cat
>         .expect_is_hungry()
>         .return_const(true);
>     assert_eq!(mock_cat.is_hungry(Duration::from_secs(1 * 3600)), false);
>     assert_eq!(mock_cat.is_hungry(Duration::from_secs(5 * 3600)), true);
>     ```
> 
> - You can use `.times(n)` to limit the number of times a mock method can be
>   called to `n` --- the mock will automatically panic when dropped if this isn't
>   satisfied.

Reviewer: ```suggestion
- Mockall has a lot more functionality, and has extensive documentation at https://docs.rs/mockall.
```

Author: I think it's important to mention just a few features (such as the one in the example). It's in the notes, so it should take up very little space in the course — but it will be useful for the participants.
